File: gamesSet/Games/Reversi.cs
using gamesSet.Models;
using gamesSet.Utils;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using Newtonsoft.Json;
using System.Numerics;

namespace gamesSet.Games
{
    public static class Reversi
    {
        private static UtilityLogic utilityLogic = new UtilityLogic();
        public static void InitGameSpecificParam(Dictionary<string, object> gameParams, GameSession gameSession)
        {
            gameParams["playerWithWhites"] = gameSession.UserCreator;

            var state = new ReversiState();

            var gameState = JsonConvert.SerializeObject(state);
            gameSession.GameState = gameState;
        }
        public static List<List<int>> tryMove(ReversiState state, int row, int col, ReversiColors player)
        {
            var flippedCells = new List<List<int>>();
            if (state.Board[row, col] != ReversiColors.Free)
            {
                return flippedCells;
            }

            //Check in all 8 directions from the clicked cell for any opponent pieces
            var directions = new List<List<int>>() {
                new List<int>(){-1, -1 },
                new List<int>(){-1, 0 },
                new List<int>(){-1, 1 },
                new List<int>(){ 0, -1 },
                new List<int>(){0, 1 },
                new List<int>(){ 1, -1 },
                new List<int>(){1, 0 },
                new List<int>(){1, 1 }
            };
            foreach (var dir in directions)
            {
                var r = row + dir[0];
                var c = col + dir[1];
                var tempFlippedCells = new List<List<int>>(); ;
                //Check if the adjacent cell is an opponent piece
                while (
                    r >= 0 && r < 8 &&
                c >= 0 && c < 8 &&
                    state.Board[r, c] != ReversiColors.Free &&
                    state.Board[r, c] != player
                )
                {
                    tempFlippedCells.Add(new List<int>() { r, c });
                    r += dir[0];
                    c += dir[1];
                }
                //If are opponent pieces in this direction that can be flipped
                if (
                r >= 0 && r < 8 &&
                    c >= 0 && c < 8 &&
                    state.Board[r, c] == player
                )
                {
                    flippedCells.AddRange(tempFlippedCells);
                }
            }
            return flippedCells;
        }

        public static bool arePossibleMovesAvailable(ReversiState state, ReversiColors player)
        {
            for (int row = 0; row < state.Board.GetLength(0); row++)
            {
                for (int col = 0; col < state.Board.GetLength(1); col++)
                {
                    var newMoves = tryMove(state, row, col, player);
                    if (newMoves.Count != 0)
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        public static int countBoardColorCells(ReversiColors[,] board, ReversiColors color)
        {
            int result = 0;
            for (int i = 0; i < board.GetLength(0); i++)
            {
                for (int j = 0; j < board.GetLength(1); j++)
                {
                    if (board[i, j] == color)
                    {
                        result++;
                    }
                }
            }
            return result;
        }

        public static void processMove(GameSession gameSession, string userName, int move)
        {
            var gameParams = JsonConvert.DeserializeObject<Dictionary<string, object>>(gameSession.GameParams);
            var playerWhites = (string)gameParams["playerWithWhites"];
            var playerBlacks = gameSession.UserCreator == playerWhites ? gameSession.SecondUser : gameSession.UserCreator;
            ReversiColors player = playerBlacks == userName ? ReversiColors.Black : ReversiColors.White;

            var stateJson = gameSession.GameState;
            var state = JsonConvert.DeserializeObject<ReversiState>(stateJson);
            int row = move / state.Board.GetLength(0);
            int col = move % state.Board.GetLength(0);
            if (state.Board[row, col] != ReversiColors.Free)
            {
                return;
            }

            var flippedCells = tryMove(state, row, col, player);

            //move is invalid
            if (flippedCells.Count == 0)
            {
                return;
            }

            //Update the board
            for (int i = 0; i < flippedCells.Count; i++)
            {
                var indexesToFlip = flippedCells[i];
                state.Board[indexesToFlip[0], indexesToFlip[1]] = player;
            }
            state.Board[row, col] = player;

            var oldPlayer = player;
            var oppositePlayer = player == ReversiColors.Black ? ReversiColors.White : ReversiColors.Black;
            var nextPlayer = oppositePlayer;

            var canContinue = true;
            if (!arePossibleMovesAvailable(state, oppositePlayer))
            {
                if (arePossibleMovesAvailable(state, oldPlayer))
                {
                    nextPlayer = oldPlayer;
                    state.AdditionalMessage = "Move goes to " + oldPlayer + " again, because " + oppositePlayer + "can't move";
                }
                else
                {
                    canContinue = false;
                    if (countBoardColorCells(state.Board, ReversiColors.Free) != 0)
                    {
                        state.AdditionalMessage = "No one player can move, end of the game";
                    }
                }
            }
            else
            {
                nextPlayer = oppositePlayer;
                if (state.AdditionalMessage != "")
                {
                    state.AdditionalMessage = "";
                }
            }

            if (nextPlayer == oppositePlayer)
            {
                gameSession.NextMoveForUser = nextPlayer == ReversiColors.White ? playerWhites : playerBlacks;
            }

            //search winner
            var numEmptyCells = countBoardColorCells(state.Board, ReversiColors.Free);
            var numWhiteCells = countBoardColorCells(state.Board, ReversiColors.White);
            var numBlackCells = countBoardColorCells(state.Board, ReversiColors.Black);
            var winner = "";
            if (numEmptyCells == 0 || numWhiteCells == 0 || numBlackCells == 0 || !canContinue)
            {
                //If there are no empty cells or one color has no pieces, the game is over
                if (numWhiteCells == numBlackCells)
                {
                    winner = "Draft";
                }
                else
                {
                    winner = numWhiteCells > numBlackCells ? playerWhites : playerBlacks;
                }

                gameSession.Status = SessionStatus.finished;
                gameSession.WinnerName = winner;
            }
            gameSession.GameState = JsonConvert.SerializeObject(state);
        }
    }
}

File: gamesSet/ClientApp/src/components/Reversi.js
import React, { useState, useEffect } from 'react';

import './Reversi.css';

export function Reversi() {
    const [board, setBoard] = useState(Array(8).fill(Array(8).fill(null)));
    const [player, setPlayer] = useState('black');
    const [winner, setWinner] = useState('');
    const [additionalMessage, setAdditionalMessage] = useState('');

    useEffect(() => {
        const newBoard = board.map((rowArray, row) =>
            rowArray.map((cell, col) => {
                if (
                    (row === 3 && col === 3) ||
                    (row === 4 && col === 4)
                ) {
                    return 'white';
                } else if (
                    (row === 3 && col === 4) ||
                    (row === 4 && col === 3)
                ) {
                    return 'black';
                } else {
                    return null;
                }
            })
        );
        setBoard(newBoard);
    }, []);

    function tryMove(board, row, col, player) {
        if (board[row][col] !== null) {
            return [];
        }

        let flippedCells = [];
        //Check in all 8 directions from the clicked cell for any opponent pieces
        const directions = [
            [-1, -1], [-1, 0], [-1, 1],
            [0, -1], [0, 1],
            [1, -1], [1, 0], [1, 1]
        ];
        for (const dir of directions) {
            let r = row + dir[0];
            let c = col + dir[1];
            let tempFlippedCells = [];
            //Check if the adjacent cell is an opponent piece
            while (
                r >= 0 && r < 8 &&
                c >= 0 && c < 8 &&
                board[r][c] !== null &&
                board[r][c] !== player
            ) {
                tempFlippedCells.push([r, c]);
                r += dir[0];
                c += dir[1];
            }
            //If are opponent pieces in this direction that can be flipped
            if (
                r >= 0 && r < 8 &&
                c >= 0 && c < 8 &&
                board[r][c] === player
            ) {
                flippedCells.push(...tempFlippedCells);
            }
        }
        //console.log('flippedCells inside')
        //console.log(flippedCells)
        //console.log(flippedCells.length)
        return flippedCells;
    }

    function arePossibleMovesAvailable(board, player) {
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                let newMoves = tryMove(board, row, col, player)
                //console.log(newMoves);
                //console.log('flippedCells length')
                //console.log(newMoves.length);
                if (newMoves.length !=0) {
                    return true;
                }
            }
        }
        return false;
    }

    function handleClick(row, col,board, setBoard, setWinner) {
        if (board[row][col] !== null) {
            return;
        }

        let flippedCells = tryMove(board, row, col, player);

        //move is invalid
        if (flippedCells.length === 0) {
            return;
        }

        //Update the board
        const newBoard = board.map((rowArray, r) =>
            rowArray.map((cell, c) =>
                (r === row && c === col) || flippedCells.some(cell => cell[0] === r && cell[1] === c)
                    ? player
                    : cell
            )
        );
        setBoard(newBoard);
        let oldPlayer = player;
        let oppositePlayer = player === 'black' ? 'white' : 'black'

        let canContinue = true;
        if (!arePossibleMovesAvailable(newBoard, oppositePlayer)) {
            if (arePossibleMovesAvailable(newBoard, oldPlayer)) {
                setPlayer(oldPlayer);
                setAdditionalMessage('Move goes to ' + oldPlayer + ' again, because ' + oppositePlayer + 'can\'t move');
            }
            else {
                canContinue = false;
                if (newBoard.flat().filter(cell => cell === null).length !== 0) {
                    setAdditionalMessage('No one player can move, end of the game');
                }
            }
        }
        else {
            setPlayer(oppositePlayer);     
            if (additionalMessage != '') {
                setAdditionalMessage('');
            }
        }

        //search winner
        const flattenedBoard = newBoard.flat();
        const numEmptyCells = flattenedBoard.filter(cell => cell === null).length;
        const numWhiteCells = flattenedBoard.filter(cell => cell === 'white').length;
        const numBlackCells = flattenedBoard.filter(cell => cell === 'black').length;
        let winner = null;
        if (numEmptyCells === 0 || numWhiteCells === 0 || numBlackCells === 0 || !canContinue) {
            //If there are no empty cells or one color has no pieces, the game is over
            if (numWhiteCells === numBlackCells) {
                winner = 'tie';
            } else {
                winner = numWhiteCells > numBlackCells ? 'white' : 'black';
            }

            setWinner(winner);
        }
    }

    return (
        <div>
            <div className="game-board">
                {board.map((row, rowIndex) => (
                    <div key={rowIndex} className="row">
                        {row.map((cell, colIndex) => (
                            <div
                                key={colIndex}
                                className={`cell ${cell}-piece`}
                                onClick={() => handleClick(rowIndex, colIndex, board, setBoard, setWinner)}
                            />
                        ))}
                    </div>
                ))}
            </div>
            <div className="status">
                {winner ? (
                    winner === 'tie' ? (
                        <p>It's a tie!</p>
                    ) : (
                        <p>{winner === 'black' ? 'Black' : 'White'} wins!</p>
                    )
                ) : (
                        <p>{player === 'black' ? 'Black' : 'White'}'s turn</p>
                )}
            </div>
            <div>{additionalMessage}</div>
        </div>
    );
}